src/ApplicationRecordLog.hpp

```diff
@@ -0,0 +1,257 @@
+            ///
+            /// This method returns the value to use when sizing the
+            /// SharedMemory object used to construct the
+            /// ApplicationRecordLog.
+            ///
+            /// @return Size requirement for SharedMemory object.
+            static size_t buffer_size(void);
+            /// @brief Gets the maximum number of records.
+            ///
+            /// This method returns the value to use when reserving
+            /// elements in the records vector passed to dump().
+            ///
+            /// @return The maximum length of the records vector after
+            ///         a call to dump().
+            static size_t max_record(void);
+            /// @brief Gets the maximum number of short region events.
+            ///
+            /// This method returns the value to use when reserving
+            /// elements in the short_regions vector passed to dump().
+            ///
+            /// @return The maximum length of the short_regions vector
+            ///         after a call to dump().
+            static size_t max_region(void);
+        protected:
+            ApplicationRecordLog() = default;
+            static constexpr size_t M_LAYOUT_SIZE = 49192;
+            static constexpr int M_MAX_RECORD = 1024;
+            static constexpr int M_MAX_REGION = M_MAX_RECORD + 1;
+    };
+    class ApplicationRecordLogImp : public ApplicationRecordLog
+    {
+        public:
+            ApplicationRecordLogImp(std::shared_ptr<SharedMemory> shmem);
+            virtual ~ApplicationRecordLogImp() = default;
+            void set_process(int process) override;
+            void set_time_zero(const geopm_time_s &time) override;
+            void enter(uint64_t hash, const geopm_time_s &time) override;
+            void exit(uint64_t hash, const geopm_time_s &time) override;
+            void epoch(const geopm_time_s &time) override;
+            void dump(std::vector<record_s> &records,
+                      std::vector<short_region_s> &short_regions) override;
+        private:
+            struct m_layout_s {
+                int num_record;
+                record_s record_table[M_MAX_RECORD];
+                int num_region;
+                short_region_s region_table[M_MAX_REGION];
+            };
+            static_assert(sizeof(m_layout_s) == M_LAYOUT_SIZE,
+                          "Defined layout size does not match the actual layout size");
+
+            struct m_region_enter_s {
+                int record_idx;
+                int region_idx;
+                geopm_time_s enter_time;
+            };
+            void check_setup(void);
+            void check_reset(m_layout_s &layout);
+            void append_record(m_layout_s &layout, const record_s &record);
+            int m_process;
+            std::shared_ptr<SharedMemory> m_shmem;
+            std::map<uint64_t, m_region_enter_s> m_hash_region_enter_map;
```

Comment:
I think you previously mentioned considering unordered_map, but that maybe there were some memory size concerns about it. Is that to-be-evaluated, or did it work out in favor of map? Or was that for a different change set?